xbmc/platform/linux/input/LibInputKeyboard.cpp

```diff
@@ -421,3 +549,55 @@ void CLibInputKeyboard::GetRepeat(libinput_device *dev)
     m_repeatData.insert(std::make_pair(dev, kbdrepvec));
   }
 }
+
+bool CLibInputKeyboard::SupportsKeyComposition() const
+{
+  return m_composedState != nullptr;
+}
+
+void CLibInputKeyboard::NotifyKeyComposingEvent(uint8_t eventType, std::uint16_t unicodeCodepoint)
+{
+  XBMC_Event event{};
+  event.type = eventType;
+  event.key.keysym.unicode = unicodeCodepoint;
+  std::shared_ptr<CAppInboundProtocol> appPort = CServiceBroker::GetAppPort();
+  if (appPort)
+  {
+    appPort->OnEvent(event);
+  }
+}
+
+std::uint32_t CLibInputKeyboard::UnicodeCodepointForKeycode(xkb_keycode_t code) const
+{
+  uint32_t unicode;
+
+  if (SupportsKeyComposition())
+  {
+    const xkb_compose_status composerStatus = xkb_compose_state_get_status(m_composedState.get());
+    if (composerStatus == XKB_COMPOSE_COMPOSED)
+    {
+      const uint32_t keysym = xkb_compose_state_get_one_sym(m_composedState.get());
+      unicode = xkb_keysym_to_utf32(keysym);
+    }
+    else
+    {
+      unicode = xkb_state_key_get_utf32(m_state.get(), code);
```

Comment:
I don't think it's possible to get here.

Reply:
Ok, maybe I'm just unfamiliar. It looked like in the switch case it returned early but I guess that only happens when backspace is pressed. So in the case here the key (other than backspace) get registered?

Reply:
It returns early if the state is composing (if the key you hit is part of a sequence). if you hit a key that cancels the composition (e.g. you hit `p` after `~` which doesn't lead to a valid sequence) you get here.
That is sent to GUI so that the buffer (`~`) is flushed and `p` fallsthrough - you end up with `~p`. Backspace is a special case, if you have `~` on the buffer and hit backspace you want to cancel the sequence so nothing is flushed on that case.